### PR TITLE
Fix a segmentation fault related to assumed length character functions.

### PR DESCRIPTION
**Pull request by qiaozhang-hnc**
When an assumed length character function is not external,  its symbol type will at first be set to `ST_IDENT` and the `AUTOBJ` field will be set to `1`. When a invocation of it is encountered, its symbol type will be modified to `ST_PROC`. But its `AUTOBJ` 
remains `1` which will causes a segmentation fault later.

This patch fixes the issue by removing the function from the automatic data list and setting its `AUTOBJ` to `0`.

The other issue is that flang sets the return length as `0` for assumed length character functions, causing an ICE afterwards.

This patch fixes it by modifying the ast type of the function (from ‘A_FUNC’ to `A_CALL`), and adding code to process the argument for the length.

### CODE REVIEW

**Comment by alokkrsharma** [comment]
Thanks for working on this. You should get a failure once you re-base it (#1157). Below change should fix that. Can I request you to include below change in your patch ?

```
--- a/tools/flang2/flang2exe/lldebug.cpp
+++ b/tools/flang2/flang2exe/lldebug.cpp
@@ -2822,7 +2822,6 @@ lldbg_create_deferred_len_string_type_mdnode(LL_DebugInfo *db, SPTR sptr,
   if (SDSCG(REVMIDLNKG(sptr)) && (DTY(DTYPEG(sptr)) == TY_PTR)) {
     /* get the array descriptor */
     SPTR sdscsptr = SDSCG(REVMIDLNKG(sptr));
-    LL_Type *dataloctype = LLTYPE(sdscsptr);
     BLKINFO *blk_info = get_lexical_block_info(db, sdscsptr, true);
     LL_MDRef file_mdnode;
     if (ll_feature_debug_info_need_file_descriptions(&db->module->ir))
@@ -2850,7 +2849,7 @@ lldbg_create_deferred_len_string_type_mdnode(LL_DebugInfo *db, SPTR sptr,

     /* emit an @llvm.dbg.declare with required !DIExpression */
     LL_MDRef expr_mdnode = lldbg_emit_expression_mdnode(db, 2, add, v1);
-    insert_llvm_dbg_declare(mdLen, sdscsptr, dataloctype,
+    insert_llvm_dbg_declare(mdLen, sdscsptr, (LL_Type *) NULL,
                             make_mdref_op(expr_mdnode), OPF_NONE);

     mdLenExp = lldbg_emit_empty_expression_mdnode(db);
```

**Comment by qiaozhang-hnc** [comment]
> Thanks for working on this. You should get a failure once you re-base it (#1157). Below change should fix that. Can I request you to include below change in your patch ?
> 
> ```
> --- a/tools/flang2/flang2exe/lldebug.cpp
> +++ b/tools/flang2/flang2exe/lldebug.cpp
> @@ -2822,7 +2822,6 @@ lldbg_create_deferred_len_string_type_mdnode(LL_DebugInfo *db, SPTR sptr,
>    if (SDSCG(REVMIDLNKG(sptr)) && (DTY(DTYPEG(sptr)) == TY_PTR)) {
>      /* get the array descriptor */
>      SPTR sdscsptr = SDSCG(REVMIDLNKG(sptr));
> -    LL_Type *dataloctype = LLTYPE(sdscsptr);
>      BLKINFO *blk_info = get_lexical_block_info(db, sdscsptr, true);
>      LL_MDRef file_mdnode;
>      if (ll_feature_debug_info_need_file_descriptions(&db->module->ir))
> @@ -2850,7 +2849,7 @@ lldbg_create_deferred_len_string_type_mdnode(LL_DebugInfo *db, SPTR sptr,
> 
>      /* emit an @llvm.dbg.declare with required !DIExpression */
>      LL_MDRef expr_mdnode = lldbg_emit_expression_mdnode(db, 2, add, v1);
> -    insert_llvm_dbg_declare(mdLen, sdscsptr, dataloctype,
> +    insert_llvm_dbg_declare(mdLen, sdscsptr, (LL_Type *) NULL,
>                              make_mdref_op(expr_mdnode), OPF_NONE);
> 
>      mdLenExp = lldbg_emit_empty_expression_mdnode(db);
> ```

OK, already added.

File: tools/flang1/flang1exe/semfunc.c [hunk 1]
```diff
@@ -1357,8 +1389,17 @@ func_call2(SST *stktop, ITEM *list, int flag)
     }
     dtype = DTYPEG(mem);
   }
-  ast = mk_func_node(A_FUNC, (callee) ? callee : mk_id(func_sptr), argt_count,
-                     argt);
+  /* As for assumed-length character functioin, use A_CALL instead of A_FUNC */
```
Comment (bryanpkc) [inline tools/flang1/flang1exe/semfunc.c:1392]: There is a typo here. I suggest rewriting this line to:
```c
/* Use A_CALL instead of A_FUNC for an assumed-length character function. */
```

File: tools/flang1/flang1exe/semfunc.c [hunk 2]
```diff
@@ -1247,8 +1248,39 @@ func_call2(SST *stktop, ITEM *list, int flag)
     goto exit_;
   }
   ii = 0;
-  /* before processing arguments, add derived type return values if needed */
-  argt = mk_argt(argt_count); /* mk_argt stuffs away count */
+  /* As for assumed-length character function, get a temporary with DT_DEFERCHAR, SC_BASED here */
```
Comment (bryanpkc) [inline tools/flang1/flang1exe/semfunc.c:1251]: I suggest rewriting this to:
```c
/* An assumed-length character function is handled as if it is a subroutine.
   Implicit temporary variables are created for the return value and its
   length, and passed to the function as output arguments. */
```

File: tools/flang1/flang1exe/semfunc.c [hunk 3]
```diff
@@ -1247,8 +1248,39 @@ func_call2(SST *stktop, ITEM *list, int flag)
...
+    add_auto_dealloc(functmp);
+    /* add the temporary and length of character to ARGT list */
+    return_value = funcast;
+    argt_count = argt_count + 2;
+    argt = mk_argt(argt_count);
+    ARGT_ARG(argt, 0) = return_value;
+    ARGT_ARG(argt, 1) = mk_convert(len_ast, stb.user.dt_int);
+    ii = ii + 2;
+  } else {
+    /* before processing arguments, add derived type return values if needed */
```
Comment (bryanpkc) [inline tools/flang1/flang1exe/semfunc.c:1281]: Is this comment still valid? This block doesn't seem to handle derived type return values.

File: tools/flang1/flang1exe/semfunc.c [hunk 4]
```diff
@@ -1247,8 +1248,39 @@ func_call2(SST *stktop, ITEM *list, int flag)
...
-  /* before processing arguments, add derived type return values if needed */
-  argt = mk_argt(argt_count); /* mk_argt stuffs away count */
+  /* As for assumed-length character function, get a temporary with DT_DEFERCHAR, SC_BASED here */
+  if ((DTY(dtype) == TY_CHAR) && ADJLENG(func_sptr)) {
+    int functmp;
+    int funcast, alloc_ast, len_ast;
+    int alloc_stmt, len_stmt;
+
+    functmp = getcctmp_sc('d', sem.dtemps++, ST_VAR, DT_DEFERCHAR, SC_BASED);
+    funcast = mk_id(functmp);
```
Comment (bryanpkc) [inline tools/flang1/flang1exe/semfunc.c:1258]: `funcast` should be renamed to `functmp_ast`, to indicate their relationship clearly.

File: tools/flang2/flang2exe/lldebug.cpp [hunk 5]
```diff
@@ -2850,7 +2849,7 @@ lldbg_create_deferred_len_string_type_mdnode(LL_DebugInfo *db, SPTR sptr,
 
     /* emit an @llvm.dbg.declare with required !DIExpression */
     LL_MDRef expr_mdnode = lldbg_emit_expression_mdnode(db, 2, add, v1);
-    insert_llvm_dbg_declare(mdLen, sdscsptr, dataloctype,
+    insert_llvm_dbg_declare(mdLen, sdscsptr, (LL_Type *) NULL,
```
Comment (bryanpkc) [inline tools/flang2/flang2exe/lldebug.cpp:2852]: Why is this change needed?

Comment (qiaozhang-hnc) [reply tools/flang2/flang2exe/lldebug.cpp:2852]: > Why is this change needed?

As @alokkrsharma suggested above, we added this change to fix the failure caused by another PR #1157

Comment (bryanpkc) [reply tools/flang2/flang2exe/lldebug.cpp:2852]: Ah, my bad, I missed that.

File: tools/flang1/flang1exe/semfunc.c (hunk 3 above)
Comment (qiaozhang-hnc) [reply tools/flang1/flang1exe/semfunc.c:1281]: > Is this comment still valid? This block doesn't seem to handle derived type return values.

Indeed, this block does not handle derived type return values, I think this comment was added as a hint that if a derived type return value exists, it should be added to the ARGT list.

**Comment by qiaozhang-hnc** [comment]
@bryanpkc This PR has been updated.

**Review by bryanpkc — APPROVED** [review]

**Review by pawosm-arm — APPROVED** [review]

**Review by shivaramaarao — APPROVED** [review]

**Comment by bryanpkc** [comment]
@qiaozhang-hnc Could you rebase this branch again?

**Comment by qiaozhang-hnc** [comment]
> @qiaozhang-hnc Could you rebase this branch again?

Rebased.
